client/my-sites/checkout/composite-checkout/wpcom/types/wpcom-store-state.ts

```diff
@@ -1,3 +1,210 @@
+type ManagedContactDetailsShape< T > = {
+	firstName: T;
+	lastName: T;
+	organization: T;
+	email: T;
+	alternateEmail: T;
+	phone: T;
+	phoneNumberCountry: T;
+	address1: T;
+	address2: T;
+	city: T;
+	state: T;
+	postalCode: T;
+	countryCode: T;
+	fax: T;
+	vatId: T;
+	tldExtraFields: ManagedContactDetailsTldExtraFieldsShape< T >;
+};
+
+type ManagedContactDetailsTldExtraFieldsShape< T > = {
+	ca?: {
+		lang: T;
+		legalType: T;
+		ciraAgreementAccepted: T;
+	};
+	uk?: {
+		registrantType: T;
+		registrationNumber: T;
+		tradingName: T;
+	};
+	fr?: {
+		registrantType: T;
+		trademarkNumber: T;
+		sirenSirat: T;
+	};
+};
+
+/*
+ * Combine two ManagedContactDetailsShape<T> objects x and y using two
+ * helper functions.
+ *
+ *   f: (A,B) => B merges fields which exist in both,
+ *   g: (A) => B constructs new fields which don't exist in y, but do in x.
+ */
+function combineManagedContactDetailsShape< A, B >(
```

Comment:
Neat! This is good for me to learn about multiple generics in a function (not sure if that's the right way to word it). Still, this is the kind of complex function which feels to me like it could benefit from having some unit tests to prove what it does and to prevent against regressions in case someone touches it in the future. What do you think?

Reply:
Good idea. Writing that test exposed a discrepancy between what I wanted this function to do, and what it actually did. :)